### PR TITLE
Refactor build system

### PR DESCRIPTION
**Pull request by timmoon10**
The build system has gotten rather clunky and unintuitive, with parallel and mostly incompatible infrastructure to configure the common, PyTorch, and JAX/TensorFlow submodules. This PR cuts down on the indirection and brings the build system closer to the typical Setuptools workflow.

The biggest change is how to specify which frameworks to build support for. Previously the user had to specify `NVTE_FRAMEWORK`, with a default of PyTorch. This PR changes the default behavior to automatically detect whatever DL frameworks are installed. Using `NVTE_FRAMEWORK` is still recommended, but this should match the naive user's expectations more closely, especially if they are working within pre-framework containers.

I've also added logic so Setuptools can automatically install dependencies:
```bash
cd transformerengine
NVTE_FRAMEWORK=jax pip install .  # Installs pybind11
NVTE_FRAMEWORK=tensorflow pip install .[test]  # Installs pybind11, pytest, tensorflow_datasets
```
It doesn't quite match the Setuptools model since the dependencies are not static (they vary based on which frameworks we are building for).

To help debug downstream dependencies, I've added the TE Git commit hash to the package version so it looks something like `0.9.0dev+2ae36a1`. This can be suppressed for public releases by setting `NVTE_NO_LOCAL_VERSION` in the environment.

TODO:
- [x] Update build tests
- [x] Update documentation
- [x] Include commit hash info, possibly in version

Future work:
- Update CI builds
- Separate JAX, TensorFlow, and userbuffers into their own Setuptools extensions, possibly without CMake
- Avoid manually loading `libtransformer_engine.so` (see https://github.com/NVIDIA/TransformerEngine/issues/115)

### CODE REVIEW

**Comment by timmoon10** [comment]
@nouiz This PR is relevant to some of your build system proposals:
- https://github.com/NVIDIA/TransformerEngine/issues/156: I couldn't figure out how to get this working
- https://github.com/NVIDIA/TransformerEngine/pull/229: Passing `--debug` as a command-line arg will build with `-g`
- https://github.com/NVIDIA/TransformerEngine/pull/233: The `[test]` option installs the test dependencies for the currently enabled frameworks

**Comment by nouiz** [comment]
Any idea when this will be merged?
This is blocking https://github.com/NVIDIA/JAX-Toolbox/issues/69,
if it isn't merged soon enough, we need to merge another fix in the mean time.

**Comment by timmoon10** [comment]
It's basically ready to go, but I'm in the process of debugging a build error in PyTorch (the GitHub test gets a memory error that I don't reproduce on my own system). Hopefully this last commit fixes it, but we'll see in 45 min.

File: setup.py [hunk 1]
```diff
@@ -2,433 +2,517 @@
...
+            import tensorflow
+        except ImportError:
+            pass
+        else:
+            _frameworks.append("tensorflow")
+
+    # Special framework names
+    if "all" in _frameworks:
+        _frameworks = supported_frameworks.copy()
+    if "none" in _frameworks:
```
Comment (nouiz) [inline setup.py:241]: what it mean to have no frameworks support?
Do we support that?

File: setup.py [hunk 2]
```diff
@@ -2,433 +2,517 @@
...
+        add_unique(setup_reqs, "ninja")
+
+    # Framework-specific requirements
+    if "pytorch" in frameworks():
+        add_unique(install_reqs, ["torch", "flash-attn>=1.0.2"])
+        add_unique(test_reqs, ["numpy", "onnxruntime", "torchvision"])
+    if "jax" in frameworks():
+        if not found_pybind11():
+            add_unique(setup_reqs, "pybind11")
+        add_unique(install_reqs, ["jax", "flax"])
```
Comment (nouiz) [inline setup.py:291]: I forgot, but isn't flax an optional dependency now?
If so, we need to move it to tests.

File: transformer_engine/CMakeLists.txt [hunk 3]
```diff
@@ -28,16 +28,20 @@ include_directories(${PROJECT_SOURCE_DIR})
...
+
 option(ENABLE_JAX "Enable JAX in the building workflow." OFF)
+message(STATUS "JAX support: ${ENABLE_JAX}")
 if(ENABLE_JAX)
   find_package(pybind11 CONFIG REQUIRED)
   add_subdirectory(jax)
 endif()
 
 option(ENABLE_TENSORFLOW "Enable TensorFlow in the building workflow." OFF)
+message(STATUS "TensorFlow support: ${ENABLE_TENSORFLOW}")
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/CMakeLists.txt:44]: PyTorch isn't handled with CMake, but with a standard-ish Setuptools extension. In the future I'd like to switch JAX and TensorFlow to that approach.

File: setup.py (hunk 1 above)
Comment (timmoon10) [reply setup.py:241]: Building without frameworks means to just compile `libtransformer_engine.so`, which might be useful for some pure C++ applications. Doing it within Setuptools might be a bit weird though.

File: setup.py (hunk 2 above)
Comment (timmoon10) [reply setup.py:291]: The plan is to make Flax optional, but right now we still import the Flax modules into the `transformer_engine.jax` namespace for backwards-compatibility. I think it would be reasonable if we modified that so we skipped those imports if Flax is not detected.

File: setup.py [hunk 4]
```diff
@@ -2,433 +2,517 @@
...
+                capture_output=True,
+                cwd=root_path,
+                check=True,
+                universal_newlines=True,
+            )
+        except (CalledProcessError, OSError):
+            pass
+        else:
+            commit = output.stdout.strip()
+            version += f"+{commit}"
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply setup.py:45]: Following https://peps.python.org/pep-0440/#local-version-identifiers, everything after `+` is considered a local label. We could also do something like `<major>.<minor>.<patch>+<num commits since release>.<commit>` so that the local labels are ordered, but that gets pretty hairy.

[the remark this answers is not in the document]
Comment (timmoon10) [reply setup.py:45]: Following https://peps.python.org/pep-0440/#local-version-identifiers, everything after `+` is considered a local label. We could also do something like `<major>.<minor>.<patch>+<num commits since release>.<commit>` so that the local labels are ordered, but that gets pretty hairy.

File: setup.py (hunk 2 above)
Comment (nouiz) [reply setup.py:291]: Lets do that in another PR.

**Comment by timmoon10** [comment]
The te_ci/L0_jax_lint test passed when I reran it, so I think we can ignore the failure as spurious.

File: docs/installation.rst [hunk 5]
```diff
@@ -48,11 +45,9 @@ Execute the following command to install the latest stable version of Transforme
...
 .. code-block:: bash
 
-  # Execute one of the following commands
-  NVTE_FRAMEWORK=pytorch pip install git+https://github.com/NVIDIA/TransformerEngine.git@stable    # Build TE for PyTorch only. The default.
-  NVTE_FRAMEWORK=jax pip install git+https://github.com/NVIDIA/TransformerEngine.git@stable        # Build TE for JAX only.
-  NVTE_FRAMEWORK=tensorflow pip install git+https://github.com/NVIDIA/TransformerEngine.git@stable # Build TE for TensorFlow only.
-  NVTE_FRAMEWORK=all pip install git+https://github.com/NVIDIA/TransformerEngine.git@stable        # Build TE for all supported frameworks.
+  pip install git+https://github.com/NVIDIA/TransformerEngine.git@stable
+
+This will automatically detect any installed deep learning frameworks and build support for them. Set `NVTE_FRAMEWORK` in the environment (as a comma-separated list) to explicitly specify which frameworks to support.
```
Comment (ksivaman) [inline docs/installation.rst:50]:
```suggestion
This will automatically detect supported deep learning frameworks installed and build TransformerEngine for them. Set `NVTE_FRAMEWORK` in the environment (as a comma-separated list) to explicitly specify which frameworks to support.
```

File: docs/installation.rst [hunk 6]
```diff
@@ -67,27 +62,38 @@ Execute the following command to install the latest development build of Transfo
...
 
-  # Execute one of the following commands
-  NVTE_FRAMEWORK=pytorch pip install git+https://github.com/NVIDIA/TransformerEngine.git@main    # Build TE for PyTorch only. The default.
-  NVTE_FRAMEWORK=jax pip install git+https://github.com/NVIDIA/TransformerEngine.git@main        # Build TE for JAX only.
-  NVTE_FRAMEWORK=tensorflow pip install git+https://github.com/NVIDIA/TransformerEngine.git@main # Build TE for TensorFlow only.
-  NVTE_FRAMEWORK=all pip install git+https://github.com/NVIDIA/TransformerEngine.git@main        # Build TE for all supported frameworks.
-  
+  pip install git+https://github.com/NVIDIA/TransformerEngine.git@main
+
+This will automatically detect any installed deep learning frameworks and build support for them. Set `NVTE_FRAMEWORK` in the environment (as a comma-separated list) to explicitly specify which frameworks to support.
```
Comment (ksivaman) [inline docs/installation.rst:67]:
```suggestion
This will automatically detect supported deep learning frameworks installed and build TransformerEngine for them. Set `NVTE_FRAMEWORK` in the environment (as a comma-separated list) to explicitly specify which frameworks to support.
```

File: docs/installation.rst (hunk 5 above)
Comment (timmoon10) [reply docs/installation.rst]: What about:

This will automatically detect if any supported deep learning frameworks are installed and build Transformer Engine support for them. To explicitly specify frameworks, set the environment variable `NVTE_FRAMEWORK` to a comma-separated list.

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -22,7 +22,7 @@ jobs:
       - name: 'Build'
         run: |
           mkdir -p wheelhouse && \
-          NVTE_FRAMEWORK=pytorch pip wheel -w wheelhouse . -v
+          NVTE_FRAMEWORK=pytorch MAX_JOBS=1 pip wheel -w wheelhouse . -v
```
Comment (ptrendx) [inline .github/workflows/build.yml:25]: Why do we need that?

File: docs/installation.rst [hunk 8]
```diff
@@ -48,11 +45,9 @@ Execute the following command to install the latest stable version of Transforme
...
 .. code-block:: bash
 
-  # Execute one of the following commands
-  NVTE_FRAMEWORK=pytorch pip install git+https://github.com/NVIDIA/TransformerEngine.git@stable    # Build TE for PyTorch only. The default.
-  NVTE_FRAMEWORK=jax pip install git+https://github.com/NVIDIA/TransformerEngine.git@stable        # Build TE for JAX only.
-  NVTE_FRAMEWORK=tensorflow pip install git+https://github.com/NVIDIA/TransformerEngine.git@stable # Build TE for TensorFlow only.
-  NVTE_FRAMEWORK=all pip install git+https://github.com/NVIDIA/TransformerEngine.git@stable        # Build TE for all supported frameworks.
+  pip install git+https://github.com/NVIDIA/TransformerEngine.git@stable
+
+This will automatically detect if any supported deep learning frameworks are installed and build Transformer Engine support for them. To explicitly specify frameworks, set the environment variable `NVTE_FRAMEWORK` to a comma-separated list.
```
Comment (ptrendx) [inline docs/installation.rst:50]: Could you add an example of that list (just for clarity)?

File: docs/installation.rst [hunk 9]
```diff
@@ -67,27 +62,38 @@ Execute the following command to install the latest development build of Transfo
...
 
-  # Execute one of the following commands
-  NVTE_FRAMEWORK=pytorch pip install git+https://github.com/NVIDIA/TransformerEngine.git@main    # Build TE for PyTorch only. The default.
-  NVTE_FRAMEWORK=jax pip install git+https://github.com/NVIDIA/TransformerEngine.git@main        # Build TE for JAX only.
-  NVTE_FRAMEWORK=tensorflow pip install git+https://github.com/NVIDIA/TransformerEngine.git@main # Build TE for TensorFlow only.
-  NVTE_FRAMEWORK=all pip install git+https://github.com/NVIDIA/TransformerEngine.git@main        # Build TE for all supported frameworks.
-  
+  pip install git+https://github.com/NVIDIA/TransformerEngine.git@main
+
+This will automatically detect if any supported deep learning frameworks are installed and build Transformer Engine support for them. To explicitly specify frameworks, set the environment variable `NVTE_FRAMEWORK` to a comma-separated list.
```
Comment (ptrendx) [inline docs/installation.rst:67]: Same.

**Review by ptrendx — APPROVED** [review]
Really cool, thanks :-).

**Review by ksivaman — APPROVED** [review]

File: .github/workflows/build.yml (hunk 7 above)
Comment (timmoon10) [reply .github/workflows/build.yml:25]: It turns out the testing systems that GitHub gives us are wimpy and experience memory issues if we let Ninja go full-force. It seems PyTorch has experienced this kind of problem before:
https://github.com/pytorch/pytorch/blob/c27cefccd371ec451a9be51fc5aced1125791493/torch/utils/cpp_extension.py#L448-L451
